File: src/main/java/cn/solarmoon/idyllic_food_diary/feature/generic_recipe/chopping/ChoppingRecipe.java
package cn.solarmoon.idyllic_food_diary.feature.generic_recipe.chopping;

import cn.solarmoon.idyllic_food_diary.registry.common.IMRecipes;
import cn.solarmoon.solarmoon_core.api.data.SerializeHelper;
import cn.solarmoon.solarmoon_core.api.entry.common.RecipeEntry;
import cn.solarmoon.solarmoon_core.api.recipe.ChanceResult;
import cn.solarmoon.solarmoon_core.api.recipe.IConcreteRecipe;
import com.google.gson.JsonObject;
import net.minecraft.core.NonNullList;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.crafting.Ingredient;
import net.minecraft.world.item.crafting.RecipeSerializer;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public record ChoppingRecipe(
        ResourceLocation id,
        Ingredient input,
        NonNullList<ChanceResult> chanceResults
) implements IConcreteRecipe {

    public Ingredient getInput() {
        return this.input;
    }

    public List<ItemStack> getResults() {
        return ChanceResult.getResults(chanceResults);
    }

    public List<ItemStack> getRolledResults(Player player) {
        return ChanceResult.getRolledResults(player, chanceResults);
    }

    @Override
    public RecipeEntry<?> getRecipeEntry() {
        return IMRecipes.CHOPPING;
    }

    @Override
    public @NotNull ResourceLocation getId() {
        return this.id;
    }

    public static class Serializer implements RecipeSerializer<ChoppingRecipe> {

        @Override
        public @NotNull ChoppingRecipe fromJson(@NotNull ResourceLocation recipeId, @NotNull JsonObject json) {
            Ingredient input = Ingredient.fromJson(json.get("ingredient"));
            NonNullList<ChanceResult> results = SerializeHelper.readChanceResults(json, "result");;
            return new ChoppingRecipe(recipeId, input, results);
        }

        @Nullable
        @Override
        public ChoppingRecipe fromNetwork(@NotNull ResourceLocation recipeId, @NotNull FriendlyByteBuf buffer) {
            Ingredient input = Ingredient.fromNetwork(buffer);
            NonNullList<ChanceResult> resultsIn = SerializeHelper.readChanceResults(buffer);
            return new ChoppingRecipe(recipeId, input, resultsIn);
        }

        @Override
        public void toNetwork(@NotNull FriendlyByteBuf buffer, ChoppingRecipe recipe) {
            recipe.input().toNetwork(buffer);
            SerializeHelper.writeChanceResults(buffer, recipe.chanceResults());
        }

    }

}

File: src/main/kotlin/cn/solarmoon/idyllic_food_diary/element/recipe/ChoppingRecipe.kt
package cn.solarmoon.idyllic_food_diary.element.recipe

import cn.solarmoon.idyllic_food_diary.registry.common.IFDRecipes
import cn.solarmoon.spark_core.api.data.RecipeJsonBuilder
import cn.solarmoon.spark_core.api.data.element.ChanceResult
import cn.solarmoon.spark_core.api.entry_builder.common.RecipeBuilder
import cn.solarmoon.spark_core.api.recipe.IConcreteRecipe
import cn.solarmoon.spark_core.feature.inlay.AttributeForgingRecipe
import cn.solarmoon.spark_core.registry.common.SparkRecipes
import com.mojang.serialization.MapCodec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.data.recipes.RecipeOutput
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.crafting.Ingredient
import net.minecraft.world.item.crafting.Recipe
import net.minecraft.world.item.crafting.RecipeSerializer

@JvmRecord
data class ChoppingRecipe(
    val input: Ingredient,
    val chanceResults: List<ChanceResult>
) : IConcreteRecipe {

    override val entry: RecipeBuilder.RecipeEntry<*>
        get() = IFDRecipes.CHOPPING

    val results: List<ItemStack>
        get() = ChanceResult.getResults(chanceResults)

    fun getRolledResults(player: Player): List<ItemStack> = ChanceResult.getRolledResults(player, chanceResults)

    class Serializer : RecipeSerializer<ChoppingRecipe> {
        override fun codec(): MapCodec<ChoppingRecipe> {
            return RecordCodecBuilder.mapCodec<ChoppingRecipe> { instance ->
                instance.group(
                    Ingredient.CODEC_NONEMPTY.fieldOf("ingredient").forGetter { it.input },
                    ChanceResult.LIST_CODEC.fieldOf("results").forGetter { it.chanceResults }
                ).apply(instance, ::ChoppingRecipe)
            }
        }

        override fun streamCodec(): StreamCodec<RegistryFriendlyByteBuf, ChoppingRecipe> {
            return StreamCodec.composite(
                Ingredient.CONTENTS_STREAM_CODEC, ChoppingRecipe::input,
                ChanceResult.LIST_STREAM_CODEC, ChoppingRecipe::chanceResults,
                ::ChoppingRecipe
            )
        }
    }

    class JsonBuilder(val r: () -> ChoppingRecipe): RecipeJsonBuilder() {
        override val name: ResourceLocation
            get() = BuiltInRegistries.ITEM.getKey(r.invoke().chanceResults[0].stack.item)

        override val prefix: String
            get() = "${IFDRecipes.CHOPPING.type.id.path}"

        override fun getRecipe(recipeOutput: RecipeOutput, location: ResourceLocation): Recipe<*> {
            return r.invoke()
        }
    }

}